pep-0544.txt

```diff
@@ -0,0 +1,949 @@
+      for t in things:
+          t.close()
+
+  f = open('foo.txt')
+  r = Resource()
+  close_all([f, r])  # OK!
+  close_all([1])     # Error: 'int' has no 'close' method
+
+Note that both the user-defined class ``Resource`` and the built-in
+``IO`` type (the return type of ``open()``) are considered subtypes of
+``SupportsClose``, because they provide a ``close()`` method with
+a compatible type signature.
+
+
+Protocol members
+----------------
+
+The methods that are protocol members are defined using the
+``@abstractmethod`` decorator. Such methods can have a default
+implementation that will be type checked. Method body can be left empty,
+i.e. be either only a docstring or a single ``...`` ellipsis literal.
+In this case type checkers will recognize that there is no default
+implementation, i.e. it can't be used via ``super()`` in explicit
+subclasses.
+
+If some or all parameters of protocol method are not annotated
+their types are assumed to be ``Any`` (see PEP 484). All other
+methods defined in the protocol class are considered non-protocol members.
+Example::
+
+  from typing import Protocol
+
+  class Example(Protocol):
+      @abstractmethod
+      def first(self) -> int:     # This is a protocol member
+          return 1
+
+      @abstractmethod
+      def second(self) -> int:    # This one also
+          ...
+
+      def auxiliary(self) -> str: # This is NOT a protocol member
+          x = 1
+
+Note that although formally the implicit return type of a method with
+an empty body is ``None``, type checker will not warn about above example,
+such convention mirrors how methods are defined in stub files.
+Abstract staticmethods, abstract classmethods, and abstract properties are
+equally supported.
+
+To define a protocol variable, one must use PEP 526 variable
+annotations in the class body. Attributes defined in the body of a method
+by assignment via ``self`` are considered non-protocol members. The rationale
+for this is that the protocol class implementation is often not shared by
+subtypes, so the interface should not depend on the default implementation.
+An exception from these rules is an attribute starting with double
+underscores and not ending with double underscores always considered
+non-protocol member. Examples::
+
+  from typing import Protocol, List
+
+  class Template(Protocol):
+      name: str        # This is a protocol member
+      value: int = 0   # This one too (with default)
+      __private: bool  # This is NOT a protocol member
+
+      def method(self) -> None:
+          self.temp: List[int] = [] # And this one neither
+
+To distinguish between protocol class variables and protocol instance
+variables, the special ``ClassVar`` annotation should be used as specified
+by PEP 526.
+
+
+Explicitly declaring implementation
+-----------------------------------
+
+To explicitly declare that a certain class implements the given protocols, they
+can be used as regular base classes. In this case:
```

Comment:
I'm worried this could be pretty confusing: you could have a class that explicitly implements some protocols and implicitly implements others.  I don't know of any other language that allows this.